Looking through the [first Jenkins build link from above](https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/17081/testReport/junit/Suite-k8s-1/11/K8sDatapathConfig_MonitorAggregation_Checks_that_monitor_aggregation_flags_send_notifications/) inside `cilium-test.log`, I see:

```
level=warning msg="/var/lib/cilium/bpf/lib/nodeport.h:1269:3: error: array size is negative" subsys=datapath-loader
level=warning msg="build_bug_on(!(NODEPORT_PORT_MIN     < NODEPORT_PORT_MAX));" subsys=datapath-loader
```

I wanted to understand what that meant, and learned `build_bug_on` is a macro that tries to create a compile-time error to validate the bpf program before loading:

```c
#define build_bug_on(e) ((void)sizeof(char[1 - 2*!!(e)]))

int main()
{
    build_bug_on(!(1 < 2)); // OK
    build_bug_on(!(3 < 2)); // Error
    build_bug_on(!(2 < 2)); // Error
}
```

The lines marked as errors trigger the exact error message as seen in the logs. This tells me that somehow `NODEPORT_PORT_MIN` was parsed out to be less than or equal to `NODEPORT_PORT_MAX`. It's also strange given that NodePortRange has defaults.

Not sure how this can happen given the code, but I will continue investigating by trying to re-run the E2E tests locally in Vagrant. Any tips for quicker feedback?

--- a/pkg/option/config.go
+++ b/pkg/option/config.go
@@ -16,6 +16,7 @@ package option
 
 import (
 	"bytes"
+	"errors"
 	"fmt"
 	"io/ioutil"
 	"net"
@@ -1828,22 +1829,9 @@ func (c *DaemonConfig) Populate() {
 	}
 	c.IPv6PodSubnets = subnets
 
-	nodePortRange := viper.GetStringSlice(NodePortRange)
-	if len(nodePortRange) > 0 {
-		if len(nodePortRange) != 2 {
-			log.Fatal("Unable to parse min/max port for NodePort range!")
-		}
-		c.NodePortMin, err = strconv.Atoi(nodePortRange[0])
-		if err != nil {
-			log.WithError(err).Fatal("Unable to parse min port value for NodePort range!")
-		}
-		c.NodePortMax, err = strconv.Atoi(nodePortRange[1])
-		if err != nil {
-			log.WithError(err).Fatal("Unable to parse max port value for NodePort range!")
-		}
-		if c.NodePortMax <= c.NodePortMin {
-			log.Fatal("NodePort range min port must be smaller than max port!")
-		}
+	err = c.populateNodePortRange()
+	if err != nil {
+		log.WithError(err).Fatal("NodePortRange can not be parsed.")
 	}
 
 	hostServicesProtos := viper.GetStringSlice(HostReachableServicesProtos)
@@ -1957,6 +1945,30 @@ func (c *DaemonConfig) Populate() {
 	c.SkipCRDCreation = viper.GetBool(SkipCRDCreation)
 	c.DisableCNPStatusUpdates = viper.GetBool(DisableCNPStatusUpdates)
 	c.AwsReleaseExcessIps = viper.GetBool(AwsReleaseExcessIps)
+}
+
+func (c *DaemonConfig) populateNodePortRange() error {
+	nodePortRange := viper.GetStringSlice(NodePortRange)
+	if len(nodePortRange) > 0 {
+		if len(nodePortRange) != 2 {
+			return errors.New("Unable to parse min/max port for NodePort range!")
+		}
+
+		var err error
+
+		c.NodePortMin, err = strconv.Atoi(nodePortRange[0])
+		if err != nil {
+			return fmt.Errorf("Unable to parse min port value for NodePort range: %s", err.Error())
+		}
+		c.NodePortMax, err = strconv.Atoi(nodePortRange[1])
+		if err != nil {
+			return fmt.Errorf("Unable to parse max port value for NodePort range: %s", err.Error())
+		}
+		if c.NodePortMax <= c.NodePortMin {
+			return errors.New("NodePort range min port must be smaller than max port!")
+		}
+	}
+	return nil
 }
 
 func sanitizeIntParam(paramName string, paramDefault int) int {

--- a/pkg/option/config.go
+++ b/pkg/option/config.go
@@ -1949,11 +1949,8 @@ func (c *DaemonConfig) Populate() {
 
 func (c *DaemonConfig) populateNodePortRange() error {
 	nodePortRange := viper.GetStringSlice(NodePortRange)
-	if len(nodePortRange) > 0 {
-		if len(nodePortRange) != 2 {
-			return errors.New("Unable to parse min/max port for NodePort range!")
-		}
-
+	switch len(nodePortRange) {
+	case 2:
 		var err error
 
 		c.NodePortMin, err = strconv.Atoi(nodePortRange[0])
@@ -1967,7 +1964,12 @@ func (c *DaemonConfig) populateNodePortRange() error {
 		if c.NodePortMax <= c.NodePortMin {
 			return errors.New("NodePort range min port must be smaller than max port!")
 		}
+	case 0:
+		log.Warning("NodePort range was set but is empty.")
+	default:
+		return errors.New("Unable to parse min/max port value for NodePort range!")
 	}
+
 	return nil
 }
 

--- a/pkg/option/config_test.go
+++ b/pkg/option/config_test.go
@@ -351,11 +351,12 @@ func Test_populateNodePortRange(t *testing.T) {
 			},
 		},
 		{
+			// We simply just want to warn the user in this case.
 			name: "NodePortRange passed as empty",
 			want: want{
 				wantMin: 0,
 				wantMax: 0,
-				wantErr: true,
+				wantErr: false,
 			},
 			preTestRun: func() {
 				viper.Reset()
@@ -365,10 +366,7 @@ func Test_populateNodePortRange(t *testing.T) {
 				fs := flag.NewFlagSet(NodePortRange, flag.ContinueOnError)
 				fs.StringSlice(
 					NodePortRange,
-					[]string{
-						fmt.Sprintf("%d", NodePortMinDefault),
-						fmt.Sprintf("%d", NodePortMaxDefault),
-					},
+					[]string{}, // Explicity has no defaults.
 					"")
 
 				BindEnv(NodePortRange)

--- a/pkg/option/config_test.go
+++ b/pkg/option/config_test.go
@@ -22,6 +22,7 @@ import (
 	"net"
 	"os"
 	"path/filepath"
+	"reflect"
 	"testing"
 
 	flag "github.com/spf13/pflag"
@@ -226,4 +227,173 @@ func (s *OptionSuite) TestLocalAddressExclusion(c *C) {
 	c.Assert(d.IsExcludedLocalAddress(net.ParseIP("3.3.3.1")), Equals, true)
 	c.Assert(d.IsExcludedLocalAddress(net.ParseIP("f00d::1")), Equals, true)
 	c.Assert(d.IsExcludedLocalAddress(net.ParseIP("f00d::2")), Equals, false)
+}
+
+func Test_populateNodePortRange(t *testing.T) {
+	type want struct {
+		wantMin int
+		wantMax int
+		wantErr bool
+	}
+	tests := []struct {
+		name       string
+		want       want
+		preTestRun func()
+	}{
+		{
+			name: "NodePortRange is valid",
+			want: want{
+				wantMin: 23,
+				wantMax: 24,
+				wantErr: false,
+			},
+			preTestRun: func() {
+				viper.Reset()
+				viper.Set(NodePortRange, []string{"23", "24"})
+			},
+		},
+		{
+			name: "NodePortRange not set in viper",
+			want: want{
+				wantMin: NodePortMinDefault,
+				wantMax: NodePortMaxDefault,
+				wantErr: false,
+			},
+			preTestRun: func() {
+				viper.Reset()
+
+				fs := flag.NewFlagSet(NodePortRange, flag.ContinueOnError)
+				fs.StringSlice(
+					NodePortRange,
+					[]string{
+						fmt.Sprintf("%d", NodePortMinDefault),
+						fmt.Sprintf("%d", NodePortMaxDefault),
+					},
+					"")
+
+				BindEnv(NodePortRange)
+				viper.BindPFlags(fs)
+			},
+		},
+		{
+			name: "NodePortMin greater than NodePortMax",
+			want: want{
+				wantMin: 666,
+				wantMax: 555,
+				wantErr: true,
+			},
+			preTestRun: func() {
+				viper.Reset()
+				viper.Set(NodePortRange, []string{"666", "555"})
+			},
+		},
+		{
+			name: "NodePortMin equal NodePortMax",
+			want: want{
+				wantMin: 666,
+				wantMax: 666,
+				wantErr: true,
+			},
+			preTestRun: func() {
+				viper.Reset()
+				viper.Set(NodePortRange, []string{"666", "666"})
+			},
+		},
+		{
+			name: "NodePortMin not a number",
+			want: want{
+				wantMin: 0,
+				wantMax: 0,
+				wantErr: true,
+			},
+			preTestRun: func() {
+				viper.Reset()
+				viper.Set(NodePortRange, []string{"aaa", "0"})
+			},
+		},
+		{
+			name: "NodePortMax not a number",
+			want: want{
+				wantMin: 1024,
+				wantMax: 0,
+				wantErr: true,
+			},
+			preTestRun: func() {
+				viper.Reset()
+				viper.Set(NodePortRange, []string{"1024", "aaa"})
+			},
+		},
+		{
+			name: "NodePortRange slice length not equal 2",
+			want: want{
+				wantMin: 0,
+				wantMax: 0,
+				wantErr: true,
+			},
+			preTestRun: func() {
+				viper.Reset()
+
+				delete(RegisteredOptions, NodePortRange)
+
+				fs := flag.NewFlagSet(NodePortRange, flag.ContinueOnError)
+				fs.StringSlice(
+					NodePortRange,
+					[]string{
+						fmt.Sprintf("%d", NodePortMinDefault),
+						fmt.Sprintf("%d", NodePortMaxDefault),
+					},
+					"")
+
+				BindEnv(NodePortRange)
+				viper.BindPFlags(fs)
+
+				viper.Set(NodePortRange, []string{"1024"})
+			},
+		},
+		{
+			name: "NodePortRange passed as empty",
+			want: want{
+				wantMin: 0,
+				wantMax: 0,
+				wantErr: true,
+			},
+			preTestRun: func() {
+				viper.Reset()
+
+				delete(RegisteredOptions, NodePortRange)
+
+				fs := flag.NewFlagSet(NodePortRange, flag.ContinueOnError)
+				fs.StringSlice(
+					NodePortRange,
+					[]string{
+						fmt.Sprintf("%d", NodePortMinDefault),
+						fmt.Sprintf("%d", NodePortMaxDefault),
+					},
+					"")
+
+				BindEnv(NodePortRange)
+				viper.BindPFlags(fs)
+
+				viper.Set(NodePortRange, []string{})
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			tt.preTestRun()
+
+			d := &DaemonConfig{}
+			err := d.populateNodePortRange()
+
+			got := want{
+				wantMin: d.NodePortMin,
+				wantMax: d.NodePortMax,
+				wantErr: err != nil,
+			}
+
+			if !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("DaemonConfig.populateNodePortRange = got %v, want %v", got, tt.want)
+			}
+		})
+	}
 }

